### ISSUE TITLE
Support for Android 4.4 FDE 

### ISSUE THREAD

**Issue by GrepItAll** [hashcat/hashcat#1442]
Although Android 4.4 has been released for 4 years now, approximately 16% of Android devices still run this OS version (as of August 2017). KitKat (4.4) has also continued to receive Security Patches as recently as September 2017.

As the last major release of Android before hardware backed FDE started to become common, it would be great if support for Android 4.4 FDE could be added into Hashcat.

This was referenced in a Twitter conversation with Nikolay Elenkov in 2014: https://twitter.com/kapitanpetko/status/484653251355082754

The python script that he referred to (https://github.com/nelenkov/Santoku-Linux/blob/master/tools/android/android_bruteforce_stdcrypto/bruteforce_stdcrypto.py) has been updated since that Twitter discussion to check magic values for ext4, as well as additional encryption checks.

Devices running 4.4 FDE continue to frustrate many people, so it makes perfect sense for the world's fastest password cracker to implement support for this ;)

I can provide sample data if needed.

Thanks in advance!

**Comment by philsmd** [comment]
Is this a duplicate of https://github.com/hashcat/hashcat/issues/411 ?

**Comment by GrepItAll** [comment]
No, although key elements are similar. #411 deals with FDE on devices where hardware backed FDE is enabled, but where the hardware keys are leaked out of the processor in the case of Qualcomm devices. 4.4 FDE does not use hardware backing. Both use scrypt however.

**Comment by jsteube** [comment]
I see this is PIN only. For only PIN, it should be fast enough on CPU. What am I missing here?

**Comment by GrepItAll** [comment]
You are correct that this PoC code is only for PINs. However, Android 4.4 lockscreens can have a password set instead, which is then used as the basis for the FDE encryption. So whilst most users would have PINs for which this script is sufficiently fast, it is not fast enough to use for passwords which are still frequent enough to cause problems when attempting to recover data from these devices.

**Comment by brandoncasaba** [comment]
To be clear, the PIN 1234 is the same as the password 1234 in the implementation. The password is simply treated as a string.

**Comment by GrepItAll** [comment]
Correct, apologies for the ambiguity. I meant that the device password can be non-numeric, whilst the python script will only handle numeric codes.
